packages/block-editor/src/components/provider/index.js

```diff
@@ -90,9 +96,19 @@ class BlockEditorProvider extends Component {
 				this.isSyncingIncomingValue = false;
 				blocks = newBlocks;
 				isPersistent = newIsPersistent;
+				lastBlockAttributesChange = newLastBlockAttributesChange;
 				return;
 			}
 
+			if ( newLastBlockAttributesChange !== lastBlockAttributesChange ) {
+				lastBlockAttributesChange = newLastBlockAttributesChange;
+
+				if ( onBlockAttributesChange ) {
+					const [ clientId, attributes ] = newLastBlockAttributesChange;
+					onBlockAttributesChange( clientId, attributes );
```

Comment:
Very first thought :) Seeing this made me wonder about an idea. Not sure yet how valuable it is or if it will allow us to improve things. but this callback could serve as a way to make the updates the blocks and "return" them.

I didn't read the PR completely but If I'm not mistaken this callback forces the caller to "set blocks" twice. When this callback is called and when `onInput` is called too?

(I might be completely wrong here, as I didn't dive yet)

Reply:
> I didn't read the PR completely but If I'm not mistaken this callback forces the caller to "set blocks" twice. When this callback is called and when `onInput` is called too?

No, because `onBlockAttributesChange` isn't actually applying anything to the blocks, it's only triggering the side effect (updating the post). As you mention, it's the subsequent `onChange` / `onInput` which calls `resetBlocks`, at which point those values are reflected in all impacted blocks.

Reply:
> No, because `onBlockAttributesChange` isn't actually applying anything to the blocks, it's only triggering the side effect (updating the post). As you mention, it's the subsequent `onChange` / `onInput` which calls `resetBlocks`, at which point those values are reflected in all impacted blocks.

An unfortunate consequence of this behavior: Since we rely on the subsequent `resetBlocks` and there's nothing about `editPost` itself which immediately triggers an application of sourced attribute values, if someone were to call `editPost` directly with updated meta values, the current implementation would not immediately update blocks which derive from this meta.

Reply:
>An unfortunate consequence of this behavior: Since we rely on the subsequent resetBlocks and there's nothing about editPost itself which immediately triggers an application of sourced attribute values, if someone were to call editPost directly with updated meta values, the current implementation would not immediately update blocks which derive from this meta.

We need a way for custom sources to subscribe to events/actions that should re-`apply` their values.